# rsRNAQuant

**rsRNAQuant** is an automated pipeline for quantification and annotation of **rRNA-derived small RNAs (rsRNAs).** It identifies and classifies rsRNAs by providing information on rsRNA length, rsRNA abundance, parental rRNA subtype, parental rRNA end, and other relevant features. It also allows users to optionally perform customized analyses of other small non-coding RNAs (sncRNAs), including tsRNAs, ysRNAs, miRNAs, and additional sncRNA types from small RNA-seq data. 

rsRNAQuant integrates preprocessing, hierarchical mapping, read assignment, normalization, and downstream visualization in a unified framework. It is implemented using **Python, Bash, R**, and **Snakemake**, ensuring reproducibility and scalability across platforms.

---

## ðŸ”¬ Overview

rsRNAQuant performs:

* Read preprocessing, including length/quality filtering, adapter trimming, and contaminant removal
* Hierarchical alignment to rRNA references
* Assignment of multi-mapping and multi-overlapping reads
* Transcript-level quantification of rsRNA sequences
* Normalization and summary statistics (RPM, length distribution, etc.)
* Visualization of rsRNA composition and expression patterns
* **Biomarker identification using machine-learningâ€“based approaches**

<img src="assets/workflow_alternative.png" alt="workflow" width="600" height="300" border="10px">

---



## âš™ï¸ Installation

### 0ï¸âƒ£ Requirements
* [[Conda](https://conda.io/)] 
* [[Snakemake](https://snakemake.readthedocs.io/en/stable/)] 

### 1ï¸âƒ£ Clone the repository
To use rsRNAQuant, you can simply **download the repository** and run the workflow out of the box. Snakemake will automatically handle all required dependencies, provided that it is installed and accessible.
```bash
git clone https://github.com/BIGRICEgo/rsRNAQuant.git
cd rsRNAQuant
```

### 2ï¸âƒ£ Create the conda environment and install dependencies

We **recommend** using **conda** together with **Mamba** for faster Snakemake installation and more efficient dependency management, preferably within a dedicated environment. The **Miniforge** distribution of conda is a good choice, as it natively supports Mamba. If conda is set up, Snakemake can be installed as follows:

```bash
conda env create -f config/environment.yml
conda activate rsrnaquant
```

If you prefer manual installation, using the command belows will install the necessary packages: 

```bash
conda create -n rsrnaquant python=3.9
conda activate rsrnaquant
pip install snakemake==9.3.3
conda install -c bioconda trim_galore bowtie samtools
conda install -c conda-forge r-base r-ggplot2 r-reshape2 r-dplyr r-tidyr r-optparse r-Hmisc r-corrplot r-jsonlite r-ggforce r-grid r-patchwork r-DESeq2 r-caret r-glmnet r-MASS r-pROC r-randomForest r-ggrepel r-RColorBrewer
pip install pandas numpy 
```

### Download annotation files & test examples

Regarding annotation references, we have pre-built the Bowtie index files. Users can simply download and decompress the package for immediate use.

1. To download annotation references, please use this link: https://yunpan.tongji.edu.cn/link/AAB740E00847784A9DA39F2EAF256F166B

    - To unpack references, please use the code belowï¼š
        ```bash
        cd annotation/
        tar -xJvf annotation.tar.xz
        ```

2. To download the test examples, please use this link: 
https://yunpan.tongji.edu.cn/link/AA71605A6F3DA24E24984C48C16704A00D

    - To unpack references, please use the code belowï¼š
      ```bash
        cd data/test/
        tar -xJvf test.tar.xz
        ```
    -  The test sample results are included in the rsRNAQuant package. You can find them in the file `Sample_test_results.tar.xz` under `rsRNAQuant/results/`, which can be extracted to view the outputs.


---

## ðŸ“‚ Directory Structure

```bash
rsRNAQuant/
â”œâ”€â”€ assets/
â”œâ”€â”€ config/
â”‚   â”œâ”€â”€ config.yaml
â”‚   â”œâ”€â”€ environment.yml
â”œâ”€â”€ annotation/
â”‚   â”œâ”€â”€ genome/
â”‚   â”œâ”€â”€ rRNAdb/
â”‚   â”œâ”€â”€ UniVec/
â”‚   â””â”€â”€ ...
â”œâ”€â”€ data/
â”‚   â”œâ”€â”€ test/
â”‚   â”‚   â”œâ”€â”€ case.fastq.gz
â”‚   â”‚   â””â”€â”€ control.fastq.gz
â”œâ”€â”€ results/
â”‚   â”œâ”€â”€ test/
â”‚   â”‚   â”œâ”€â”€ mapped/
â”‚   â”‚   â”œâ”€â”€ stats/
â”‚   â”‚   â”œâ”€â”€ plots/
â”‚   â”‚   â””â”€â”€ test_report.html
â”œâ”€â”€ scripts/
â”‚   â”œâ”€â”€ alignment_summary.sh
â”‚   â”œâ”€â”€ count_unique_sequences.sh
â”‚   â”œâ”€â”€ map_rRNA.sh
â”‚   â”œâ”€â”€ map_ncRNA.sh
â”‚   â”œâ”€â”€ re_logs.sh
â”‚   â”œâ”€â”€ process_rna_seq.py
â”‚   â”œâ”€â”€ CountMatrix.py
â”‚   â”œâ”€â”€ ExprMatrix.py
â”‚   â”œâ”€â”€ generate_report
â”‚   â”œâ”€â”€ RPM.py
â”‚   â”œâ”€â”€ compare_Plot_rRNA_coverage.R
â”‚   â”œâ”€â”€ DA.R
â”‚   â”œâ”€â”€ rRNA_length_distribution.R
â”‚   â””â”€â”€ template_report.html
â””â”€â”€ Snakefile
```

---


## ðŸ§¬ Configuration

Modify parametes inside `config/config.yaml` before running the workflow!!

  - Update your project and sample names, then assign them to the case and control groups.


Example:

```yaml
# config/config.yaml 

# Directories
raw_data_dir: "data" # Path to raw data
results_dir: "results" # Path to output directory
scripts_dir: "scripts" # Path to script directory
annotation_dir: "annotation" # Path to annotation directory
ncRNA: false # Whether to perform analysis of ncRNAs, "true" or "false"
intermediate_files: false # Whether to all mapping details, "true" or "false"

# Project infos
project: "Test" # Project name
samples: # Sample names
  - case1
  - case2
  - case3
  - control1
  - control2
  - control3
groups: # Group to which samples belong.
  case:
    - case1
    - case2
    - case3
  control:
    - control1
    - control2
    - control3
```

---

## â–¶ï¸ Running the Pipeline

âš ï¸ Before running, check that you have downloaded the annotations and already modified the configfile.

### Dry run (check DAG)

```bash
snakemake -n --cores 4
```

### ðŸ§ª Full execution

```bash
snakemake --cores 8 
```

To visualize the workflow:

```bash
snakemake --dag | dot -Tpdf > DAG.pdf
```

### Other condition

To Force rerun:
```bash
snakemake --force
```
To finish uncompleted jobs:
```bash
snakemake --rerun-incomplete
```


---

## ðŸ“Š Results

After successful execution, results will be saved under `results/{Project}/`.

| Folder     | Description                                   |
| ---------- | --------------------------------------------- |
| `mapped/`  | Alignment results for genome, rRNA, and ncRNA |
| `stats/`   | Expression matrices and distribution tables   |
| `logs/`    | Execution logs for each Sample                |
| `plots/`   | Plots and biomarker results                   |

Main outputs include:

* `*_alignment_summary_rRNA.txt`: Hierarchical rRNA assignment details
* `*_expression_matrix.txt`: Normalized sncRNA expression table
* `*_rRNA_length_distribution.txt`: Length distribution per rRNA category
* `*_DA_result.txt`: Differential analysis results
* `*_lasso_selected_features.txt`: LASSO-selected features table
* PDF/PNG visualizations: distribution barplots, category pie charts, coverage lineplots, co-expression heatmaps, feature importance dotplots, ROC curves, etc.

<img src="assets/result_plot.png" alt="result_plots" width="550" height="350" border="10px">


### Result report
The result report is generated in the `Results/` directory, and the report is a HTML file which can be opened in any Web browser.

<img src="assets/result_report_p1.png" alt="result_plots" width="550" height="700" border="10px">

---

## ðŸ§­ FAQs

**Q1.** `MissingInputException: Missing input files for rule trim_reads` 

â†’ Ensure raw FASTQ files are correctly placed in `data/` and match the sample names in `config.yaml`.

**Q2.** How to regenerate the DAG after edits?

```bash
snakemake --dag | dot -Tpdf > DAG.pdf
```

**Q3.** How to clean intermediate files?

Add `temp()` to intermediate outputs in the Snakefile or run:

```bash
snakemake --delete-temp-output
```

---

## â›µï¸ Advanced Usage

The Snakemake framework provides flexible rule definition and workflow control, allowing users to customize rules according to specific needs. For example:

1. **Optional Alignment to Other ncRNA Types**

    Users can enable analysis of additional small non-coding RNAs by setting the ncRNA parameter to true in the `config/config.yaml` file (default: false, which restricts analysis to rsRNAs only). 

2. **Retention of Intermediate Files**

    To retain intermediate alignment files (stored in `results/mapped/`), set intermediate_files to true in config/config.yaml. Users should ensure sufficient disk space is available when enabling this option.



---

## ðŸ“šï¸Ž More Infos

ðŸ‘‰ See more usage details and **full-size report.html** in [[**Guidelines**](http://rsrnadisease.zhanglab-bioinfo.cn/Tool)] for rsRNAQuant. 

ðŸš© More informations could be found in related database: [[**rsRNADisease**](http://rsrnadisease.zhanglab-bioinfo.cn/)].


---

## ðŸ‘©â€ðŸ”¬ Authors

**Huang RF**, PhD Candidate [Zhanglab / Tongji Uni]

If you have any questions, bug reports, or suggestions, please contact:

  **Prof. Xiao-Ou Zhang**  

- See our recent work: [https://xozhanglab.com/]

- Contact: [ðŸ“§ zhangxiaoou@tongji.edu.cn](zhangxiaoou@tongji.edu.cn)

---


## ðŸ§¾ Citation

If you use rsRNAQuant in your work, please cite:

> **Huang RF**, **Zhang XO**.
> **rsRNAQuant: a reproducible pipeline for comprehensive quantification of extracellular vesicle small RNAs**.
> (Manuscript in preparation)
